File: packages/python/src/callairis_event_forge/services/outbox_service.py
import asyncio
import logging
import random
from datetime import datetime, timedelta, timezone
from typing import Any, Optional

from ..config import OutboxConfig
from ..errors import ProcessingError
from ..events import OutboxEvents, SimpleEventEmitter
from ..models import CreateOutboxMessageDto, OutboxMessage
from ..publishers.interfaces import IMessagePublisher
from ..repositories.interfaces import IOutboxRepository

logger = logging.getLogger(__name__)


class OutboxService(SimpleEventEmitter):
    """Outbox Service — manages message creation, polling, and publishing.

    Port of the Node.js OutboxService with:
    - Exponential backoff with jitter
    - Stale lock release
    - Concurrency guard (isProcessing flag)
    - Event emission matching Node.js events
    - Cleanup timer for old messages
    """

    def __init__(
        self,
        repository: IOutboxRepository,
        publisher: IMessagePublisher,
        config: Optional[OutboxConfig] = None,
    ) -> None:
        super().__init__()
        self.repository = repository
        self.publisher = publisher
        self.config = config or OutboxConfig()
        self._polling_task: Optional[asyncio.Task[None]] = None
        self._cleanup_task: Optional[asyncio.Task[None]] = None
        self._is_processing = False

    async def create_message(
        self,
        dto: CreateOutboxMessageDto,
        transaction_context: Any | None = None,
    ) -> OutboxMessage:
        """Create a new outbox message, optionally within a transaction context."""
        message = await self.repository.create(dto, transaction_context)

        if self.config.immediate_processing:
            self.emit(OutboxEvents.MESSAGE_CREATED, message.id)

        return message

    async def with_transaction(self, operation: Any) -> Any:
        """Execute operation within a transaction."""
        return await self.repository.with_transaction(operation)

    def start_polling(self) -> None:
        """Start polling for pending messages."""
        if self._polling_task is not None:
            return

        self.emit(OutboxEvents.POLLING_STARTED)
        self._polling_task = asyncio.create_task(self._poll_loop())

        # Start cleanup timer
        self._start_cleanup()

        # Initial poll
        asyncio.create_task(self._poll_and_process())

    def stop_polling(self) -> None:
        """Stop polling."""
        if self._polling_task is not None:
            self._polling_task.cancel()
            self._polling_task = None
            self.emit(OutboxEvents.POLLING_STOPPED)

        if self._cleanup_task is not None:
            self._cleanup_task.cancel()
            self._cleanup_task = None

    async def process_message(self, _message_id: str) -> None:
        """Process a specific message immediately (called via event)."""
        await self._poll_and_process()

    async def _poll_loop(self) -> None:
        """Polling loop — runs until cancelled."""
        while True:
            try:
                await asyncio.sleep(self.config.polling_interval)
                await self._poll_and_process()
            except asyncio.CancelledError:
                break
            except Exception as e:
                logger.error(f"Error in outbox polling loop: {e}", exc_info=True)
                self.emit("error", e)

    async def _poll_and_process(self) -> None:
        """Fetch and process pending messages with concurrency guard."""
        if self._is_processing:
            return

        self._is_processing = True
        try:
            # Release stale locks first
            lock_timeout = datetime.now(timezone.utc) - timedelta(seconds=self.config.lock_timeout_seconds)
            await self.repository.release_stale_locks(lock_timeout)

            # Fetch and lock pending messages
            messages = await self.repository.fetch_and_lock_pending(
                self.config.batch_size, self.config.worker_id
            )

            if not messages:
                return

            # Process each message concurrently
            await asyncio.gather(
                *(self._publish_message(msg) for msg in messages),
                return_exceptions=True,
            )
        except Exception as e:
            logger.error(f"Error in poll_and_process: {e}", exc_info=True)
            self.emit("error", e)
        finally:
            self._is_processing = False

    async def _publish_message(self, message: OutboxMessage) -> None:
        """Publish a single message."""
        try:
            await self.publisher.publish(message)
            await self.repository.mark_published(message.id)
            self.emit(OutboxEvents.MESSAGE_PUBLISHED, message)
        except Exception as e:
            await self._handle_publish_error(message, e)

    async def _handle_publish_error(self, message: OutboxMessage, error: Exception) -> None:
        """Handle publishing error with retry logic and backoff."""
        error_message = str(error)
        is_permanent = isinstance(error, ProcessingError)

        if message.retry_count >= message.max_retries or is_permanent:
            await self.repository.mark_failed(message.id, error_message, permanent=True)
            self.emit(OutboxEvents.MESSAGE_FAILED, {
                "message": message, "error": error, "permanent": True,
            })
            return

        # Calculate exponential backoff with jitter
        backoff_delay = self._calculate_backoff(message.retry_count)
        scheduled_at = datetime.now(timezone.utc) + timedelta(milliseconds=backoff_delay)

        await self.repository.mark_failed(
            message.id, error_message, permanent=False, scheduled_at=scheduled_at
        )
        self.emit(OutboxEvents.MESSAGE_FAILED, {
            "message": message, "error": error, "permanent": False,
        })

    def _calculate_backoff(self, retry_count: int) -> float:
        """Calculate exponential backoff delay in milliseconds.

        Formula: min(base * 2^retryCount, max) ± 10% jitter
        """
        base = self.config.backoff_base_seconds
        max_delay = self.config.max_backoff_seconds

        exponential_delay = base * (2 ** retry_count)
        capped_delay = min(exponential_delay, max_delay)

        jitter = capped_delay * 0.1 * (random.random() * 2 - 1)
        final_delay = capped_delay + jitter

        return max(0.0, final_delay * 1000)

    def _start_cleanup(self) -> None:
        """Start cleanup timer for old messages."""
        self._cleanup_task = asyncio.create_task(self._cleanup_loop())

    async def _cleanup_loop(self) -> None:
        """Cleanup loop — runs until cancelled."""
        # Initial cleanup
        await self._cleanup()
        while True:
            try:
                await asyncio.sleep(self.config.cleanup_interval)
                await self._cleanup()
            except asyncio.CancelledError:
                break
            except Exception as e:
                logger.error(f"Error in outbox cleanup: {e}", exc_info=True)
                self.emit("error", e)

    async def _cleanup(self) -> None:
        """Clean up old published messages."""
        try:
            cutoff_date = datetime.now(timezone.utc) - timedelta(days=self.config.retention_days)
            deleted = await self.repository.delete_older_than(cutoff_date)
            if deleted > 0:
                self.emit("cleanup", {"deleted": deleted, "cutoff_date": cutoff_date})
        except Exception as e:
            self.emit("error", e)

    async def cleanup_old_messages(self, older_than_days: int = 30) -> int:
        """Manual cleanup of old messages."""
        cutoff_date = datetime.now(timezone.utc) - timedelta(days=older_than_days)
        return await self.repository.delete_older_than(cutoff_date)

File: packages/core/src/services/outbox.service.ts
import { EventEmitter } from 'events';

import { DEFAULT_OUTBOX_CONFIG, OutboxConfig } from '../config/outbox.config';
import { ProcessingError } from '../errors/processing.error';
import { CreateOutboxMessageDto } from '../interfaces/create-outbox-message.dto';
import { IMessagePublisher } from '../interfaces/message-publisher.interface';
import { OutboxMessage } from '../interfaces/outbox-message.interface';
import { IOutboxRepository } from '../interfaces/outbox-repository.interface';

/**
 * Outbox Service Events
 */
export enum OutboxEvents {
  MESSAGE_CREATED = 'outbox:message:created',
  MESSAGE_PUBLISHED = 'outbox:message:published',
  MESSAGE_FAILED = 'outbox:message:failed',
  PROCESSOR_STARTED = 'outbox:processor:started',
  PROCESSOR_STOPPED = 'outbox:processor:stopped',
  /** @deprecated Use PROCESSOR_STARTED */
  POLLING_STARTED = 'outbox:polling:started',
  /** @deprecated Use PROCESSOR_STOPPED */
  POLLING_STOPPED = 'outbox:polling:stopped',
}

/**
 * Outbox Service
 * Manages outbox message creation, event-driven publishing, and safety-net polling
 */
export class OutboxService extends EventEmitter {
  private readonly config: Required<OutboxConfig>;
  private safetyNetTimer?: NodeJS.Timeout;
  private cleanupTimer?: NodeJS.Timeout;
  private isProcessing = false;
  private retryTimers = new Map<string, NodeJS.Timeout>();

  constructor(
    private readonly repository: IOutboxRepository,
    private readonly publisher: IMessagePublisher,
    config?: OutboxConfig,
  ) {
    super();
    this.config = {
      ...DEFAULT_OUTBOX_CONFIG,
      ...config,
      safetyNetInterval: config?.safetyNetInterval ?? config?.pollingInterval ?? DEFAULT_OUTBOX_CONFIG.safetyNetInterval,
    };
  }

  /**
   * Create a new outbox message
   * Optionally within a transaction context
   */
  async createMessage(
    dto: CreateOutboxMessageDto,
    transactionContext?: unknown,
  ): Promise<OutboxMessage> {
    const message = await this.repository.create(
      {
        ...dto,
        maxRetries: dto.maxRetries ?? this.config.maxRetries,
      },
      transactionContext,
    );

    // Emit event for immediate processing
    if (this.config.immediateProcessing) {
      this.emit(OutboxEvents.MESSAGE_CREATED, message.id);
    }

    return message;
  }

  /**
   * Execute operation within a transaction
   */
  async withTransaction<T>(operation: (context: unknown) => Promise<T>): Promise<T> {
    return this.repository.withTransaction(operation);
  }

  /**
   * Start the event-driven processor
   * - Listens for MESSAGE_CREATED events for immediate publish
   * - Runs a low-frequency safety net poll for crash recovery
   */
  startProcessor(): void {
    if (this.safetyNetTimer) {
      return;
    }

    // Event-driven: immediate publish on create
    this.on(OutboxEvents.MESSAGE_CREATED, (id: string) => {
      void this.processMessageById(id);
    });

    // Safety net: low-frequency poll for crash recovery
    this.safetyNetTimer = setInterval(() => {
      void this.pollAndProcess();
    }, this.config.safetyNetInterval);

    this.startCleanup();

    this.emit(OutboxEvents.PROCESSOR_STARTED);
    this.emit(OutboxEvents.POLLING_STARTED);
  }

  /**
   * Stop the processor
   */
  stopProcessor(): void {
    if (this.safetyNetTimer) {
      clearInterval(this.safetyNetTimer);
      this.safetyNetTimer = undefined;
    }

    if (this.cleanupTimer) {
      clearInterval(this.cleanupTimer);
      this.cleanupTimer = undefined;
    }

    // Clear all retry timers
    for (const timer of this.retryTimers.values()) {
      clearTimeout(timer);
    }
    this.retryTimers.clear();

    // Remove MESSAGE_CREATED listeners added by startProcessor
    this.removeAllListeners(OutboxEvents.MESSAGE_CREATED);

    this.emit(OutboxEvents.PROCESSOR_STOPPED);
    this.emit(OutboxEvents.POLLING_STOPPED);
  }

  /**
   * @deprecated Use startProcessor() instead
   */
  startPolling(): void {
    this.startProcessor();
  }

  /**
   * @deprecated Use stopProcessor() instead
   */
  stopPolling(): void {
    this.stopProcessor();
  }

  /**
   * Process a specific message by ID (event-driven path)
   */
  async processMessageById(id: string): Promise<void> {
    try {
      const message = await this.repository.findAndLockById(id, this.config.workerId);
      if (!message) return;
      await this.publishMessage(message);
    } catch (error) {
      this.emit('error', error);
    }
  }

  /**
   * Process a specific message immediately
   * @deprecated This method polls all pending messages. Use processMessageById() instead.
   */
  async processMessage(_messageId: string): Promise<void> {
    await this.pollAndProcess();
  }

  /**
   * Poll for pending messages and process them (safety net)
   */
  private async pollAndProcess(): Promise<void> {
    if (this.isProcessing) {
      return;
    }

    this.isProcessing = true;

    try {
      // Release stale locks first
      const lockTimeout = new Date(Date.now() - this.config.lockTimeoutSeconds * 1000);
      await this.repository.releaseStaleLocks(lockTimeout);

      // Fetch and lock pending messages
      const messages = await this.repository.fetchAndLockPending(
        this.config.batchSize,
        this.config.workerId,
      );

      //Guard against undefined/null
      if (!messages || messages.length === 0) {
        return;
      }

      // Process each message
      await Promise.all(messages.map((message) => this.publishMessage(message)));
    } catch (error) {
      // Log error but don't stop polling
      this.emit('error', error);
    } finally {
      this.isProcessing = false;
    }
  }

  /**
   * Publish a single message
   */
  private async publishMessage(message: OutboxMessage): Promise<void> {
    try {
      // Publish to message broker
      await this.publisher.publish(message);

      // Mark as published
      await this.repository.markPublished(message.id);

      this.emit(OutboxEvents.MESSAGE_PUBLISHED, message);
    } catch (error) {
      await this.handlePublishError(message, error);
    }
  }

  /**
   * Handle publishing error with retry logic
   */
  private async handlePublishError(message: OutboxMessage, error: unknown): Promise<void> {
    const errorMessage = error instanceof Error ? error.message : String(error);
    const isPermanent = error instanceof ProcessingError;

    // Check if we've exceeded max retries
    if (message.retryCount >= message.maxRetries || isPermanent) {
      await this.repository.markFailed(message.id, errorMessage, true);
      this.emit(OutboxEvents.MESSAGE_FAILED, { message, error, permanent: true });
      return;
    }

    // Calculate exponential backoff with jitter
    const backoffDelay = this.calculateBackoff(message.retryCount);
    const scheduledAt = new Date(Date.now() + backoffDelay);

    // Mark as failed with scheduled retry
    await this.repository.markFailed(message.id, errorMessage, false, scheduledAt);
    this.emit(OutboxEvents.MESSAGE_FAILED, { message, error, permanent: false });

    // Schedule retry via setTimeout (event-driven retry)
    const timer = setTimeout(() => {
      this.retryTimers.delete(message.id);
      void this.processMessageById(message.id);
    }, backoffDelay);
    this.retryTimers.set(message.id, timer);
  }

  /**
   * Calculate exponential backoff delay in milliseconds
   * Formula: min(backoffBaseSeconds * 2^retryCount, maxBackoffSeconds) + jitter
   */
  private calculateBackoff(retryCount: number): number {
    const baseDelaySeconds = this.config.backoffBaseSeconds;
    const maxDelaySeconds = this.config.maxBackoffSeconds;

    // Exponential backoff: base * 2^retryCount
    const exponentialDelay = baseDelaySeconds * Math.pow(2, retryCount);

    // Apply max limit
    const cappedDelay = Math.min(exponentialDelay, maxDelaySeconds);

    // Add jitter (±10% randomization to prevent thundering herd)
    const jitter = cappedDelay * 0.1 * (Math.random() * 2 - 1);
    const finalDelay = cappedDelay + jitter;

    // Convert to milliseconds
    return Math.max(0, finalDelay * 1000);
  }

  /**
   * Start cleanup timer for old messages
   */
  private startCleanup(): void {
    this.cleanupTimer = setInterval(() => {
      void this.cleanup();
    }, this.config.cleanupInterval);

    // Initial cleanup
    void this.cleanup();
  }

  /**
   * Clean up old published messages
   */
  private async cleanup(): Promise<void> {
    try {
      const cutoffDate = new Date(Date.now() - this.config.retentionDays * 24 * 60 * 60 * 1000);
      const deleted = await this.repository.deleteOlderThan(cutoffDate);

      if (deleted > 0) {
        this.emit('cleanup', { deleted, cutoffDate });
      }
    } catch (error) {
      this.emit('error', error);
    }
  }

}
